dnscrypt-proxy/sources.go

```diff
@@ -135,7 +135,8 @@ func fetchFromURL(xTransport *XTransport, u *url.URL) (bin []byte, err error) {
 }
 
 func (source *Source) fetchWithCache(xTransport *XTransport, now time.Time) (delay time.Duration, err error) {
-	if delay, err = source.fetchFromCache(now); err != nil {
+	var errCached error
+	if delay, errCached = source.fetchFromCache(now); errCached != nil {
 		if len(source.urls) == 0 {
 			dlog.Errorf("Source [%s] cache file [%s] not present and no valid URL", source.name, source.cacheFile)
 			return
```

Comment:
This return needs to reference the new `errCached` variable explicitly otherwise it always returns a nil `err`. Should fix unit test failures for `download_no-urls` and `download_path-err`.

Reply:
I have updated the code, please review.

1. cache expired or read cache failed: return delay 0
2. pass `errCached` forward for empty urls

Reply:
The `download_no-urls` and `download_path-err` test failures are not yet resolved, the return statement on this line is still always returning `nil` error even when there is a fatal error.

Reply:
For safety, since there are other returns before you set `err`, I would revert the`if` statement and then set `errCached = err` afterwards.